packages/relay/src/lib/services/cacheService/cacheService.ts

```diff
@@ -0,0 +1,187 @@
+    this.sharedCache = this.internalCache;
+    this.isSharedCacheEnabled = this.isRedisEnabled();
+
+    if (this.isSharedCacheEnabled) {
+      this.sharedCache = new RedisCache(logger.child({ name: 'redisCache' }), register);
+    }
+  }
+
+  /**
+   * Checks whether Redis caching is enabled based on environment variables.
+   * @private
+   * @returns {boolean} Returns true if Redis caching is enabled, otherwise false.
+   */
+  private isRedisEnabled(): boolean {
+    const redisEnabled = process.env.REDIS_ENABLED && process.env.REDIS_ENABLED === 'true';
+    const redisUrlValid = process.env.REDIS_URL && process.env.REDIS_URL !== '';
+
+    if (redisEnabled && redisUrlValid) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Retrieves a cached value associated with the given key.
+   * If the shared cache is enabled and an error occurs while accessing it,
+   * the internal cache is used as a fallback.
+   * @param {string} key - The key associated with the cached value.
+   * @param {string} callingMethod - The name of the method calling the cache.
+   * @param {string} requestIdPrefix - A prefix to include in log messages (optional).
+   * @param {boolean} shared - Whether to use the shared cache (optional, default: false).
+   * @returns {*} The cached value if found, otherwise null.
+   */
+  public async get(key: string, callingMethod: string, requestIdPrefix?: string, shared: boolean = false): Promise<any> {
+    if (shared && this.isSharedCacheEnabled) {
+      try {
+        return await this.sharedCache.get(key, callingMethod, requestIdPrefix);
+      } catch (error) {
+        const redisError = new RedisCacheError(error);
+        this.logger.error(`Error occured while getting the cache from Redis. Fallback to internal cache. Error is: ${redisError.fullError}`);
+      }
+      return this.internalCache.get(key, callingMethod, requestIdPrefix);
+    } else {
+      return this.internalCache.get(key, callingMethod, requestIdPrefix);
+    }
+  }
+
+  /**
+   * Sets a value in the cache associated with the given key.
+   * If the shared cache is enabled and an error occurs while setting in it,
+   * the internal cache is used as a fallback.
+   * @param {string} key - The key to associate with the value.
+   * @param {*} value - The value to cache.
+   * @param {string} callingMethod - The name of the method calling the cache.
+   * @param {number} ttl - Time to live for the cached value in milliseconds (optional).
+   * @param {string} requestIdPrefix - A prefix to include in log messages (optional).
+   * @param {boolean} shared - Whether to use the shared cache (optional, default: false).
+   */
+  public set(
+    key: string,
+    value: any,
+    callingMethod: string,
+    ttl?: number,
+    requestIdPrefix?: string,
+    shared: boolean = false
+  ): void {
+    if (shared && this.isSharedCacheEnabled) {
+      try {
+        return this.sharedCache.set(key, value, callingMethod, ttl, requestIdPrefix);
```

Comment:
We can wait it out or just execute it and continue forward without waiting. It will be executed in the background.

Reply:
Ohhh sneaky. Hmmm
Whiles optimal in most cases it might result in some double sets because a future process gets, find nothing and then sets.
Maybe add a note as to this intent and let watch how things unfold